### PR TITLE
drivers: spi_esp32_spim: Add support for IOMUX mode

### PR DESCRIPTION
**Pull request by xingrz**
In order to work on a clock speed higher than 20 MHz, IO MUX is required.

References:
* https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32/api-reference/peripherals/spi_master.html#c.SPICOMMON_BUSFLAG_IOMUX_PINS
* https://docs.espressif.com/projects/esp-idf/en/release-v4.4/esp32/api-reference/peripherals/gpio.html#overview
* https://github.com/espressif/esp-idf/blob/ee2029c38ac05bad3ac4c4684a9c0f37094e9246/components/driver/spi_common.c#L553

### CODE REVIEW

**Comment by sylvioalves** [comment]
@xingrz, sorry for the delay. I'll check whether this information can be retrieved from pinctrl data instead/wifi of having a KConfig.

**Comment by xingrz** [comment]
> @xingrz, sorry for the delay. I'll check whether this information can be retrieved from pinctrl data instead/wifi of having a KConfig.

Retrieving from pinctrl data is possible: https://github.com/espressif/esp-idf/blob/ee2029c38ac05bad3ac4c4684a9c0f37094e9246/components/driver/spi_common.c#L373

**Review by glaubermaroto — CHANGES_REQUESTED** [review]
Make it a dts property instead of a Kconfig option [https://docs.zephyrproject.org/latest/build/kconfig/tips.html#what-not-to-turn-into-kconfig-options](https://docs.zephyrproject.org/latest/build/kconfig/tips.html#what-not-to-turn-into-kconfig-options)

**Comment by xingrz** [comment]
> Make it a dts property instead of a Kconfig option https://docs.zephyrproject.org/latest/build/kconfig/tips.html#what-not-to-turn-into-kconfig-options

Updated. Thank you!

File: dts/bindings/spi/espressif,esp32-spi.yaml [hunk 1]
```diff
@@ -55,3 +55,12 @@ properties:
...
       description: Make CS positive during a transaction instead of negative
+
+    iomux:
+      type: boolean
+      required: false
+      description: |
+        Routes SPI pins through IO MUX, instead of GPIO matrix.
+
+        This enables the SPI driver to work in a clock speed higher than 20 MHz.
+        Only available for some specific pins.
```
Comment (glaubermaroto) [inline dts/bindings/spi/espressif,esp32-spi.yaml:66]:
```suggestion
    use-iomux:
      type: boolean
      required: false
      description: |
        Some pins are allowed to bypass the GPIO Matrix and use the IO_MUX
        routing mechanism instead, this avoids extra routing latency and makes
        possible the use of operating frequencies higher than 20 MHz.

        Refer to SoC's Technical Reference Manual to check which pins are
        allowed to use this routing path.
```

File: drivers/spi/spi_esp32_spim.h [hunk 2]
```diff
@@ -28,6 +28,7 @@ struct spi_esp32_config {
 	int irq_source;
 	const struct pinctrl_dev_config *pcfg;
 	clock_control_subsys_t clock_subsys;
+	bool iomux;
```
Comment (glaubermaroto) [inline drivers/spi/spi_esp32_spim.h:31]:
```suggestion
	bool use_iomux;
```

File: drivers/spi/spi_esp32_spim.c [hunk 3]
```diff
@@ -365,6 +365,7 @@ static const struct spi_driver_api spi_api = {
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),	\
 		.clock_subsys =	\
 			(clock_control_subsys_t)DT_INST_CLOCKS_CELL(idx, offset),	\
+		.iomux = DT_INST_PROP(idx, iomux),	\
```
Comment (glaubermaroto) [inline drivers/spi/spi_esp32_spim.c:368]:
```suggestion
		.iomux = DT_INST_PROP(idx, use_iomux),	\
```

File: drivers/spi/spi_esp32_spim.c [hunk 4]
```diff
@@ -199,7 +199,7 @@ static int IRAM_ATTR spi_esp32_configure(const struct device *dev,
 		.clock_speed_hz = spi_cfg->frequency,
 		.duty_cycle = cfg->duty_cycle == 0 ? 128 : cfg->duty_cycle,
 		.input_delay_ns = cfg->input_delay_ns,
-		.use_gpio = true
+		.use_gpio = !cfg->iomux,
```
Comment (glaubermaroto) [inline drivers/spi/spi_esp32_spim.c:202]:
```suggestion
		.use_gpio = !cfg->use_iomux,
```

**Comment by glaubermaroto** [comment]
@xingrz Please check the compliance log messages. The commit signature needs a fix.

**Review by glaubermaroto — DISMISSED** [review]

**Comment by xingrz** [comment]
@glaubermaroto Updated.

**Review by sylvioalves — DISMISSED** [review]

**Review by uLipe — DISMISSED** [review]

**Review by glaubermaroto — APPROVED** [review]

**Review by LucasTambor — APPROVED** [review]

**Pull request merged**